File: leetcode/binary-tree-inorder-traversal/binary-tree-inorder-traversal-iter.rb
# LeetCode #94 - Binary Tree Inorder Traversal
# https://leetcode.com/problems/binary-tree-inorder-traversal/
# Iterative solution.

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def inorder_traversal(root)
  out = []
  stack = []

  while root || !stack.empty?
    # Go left as far as possible.
    while root
      stack << root
      root = root.left
    end

    # Visit the current node and prepare to go right.
    cur = stack.pop
    out << cur.val
    root = cur.right
  end

  out
end

File: leetcode/binary-tree-inorder-traversal/binary-tree-inorder-traversal-iter.java
// LeetCode #94 - Binary Tree Inorder Traversal
// https://leetcode.com/problems/binary-tree-inorder-traversal/
// Iterative implementation.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> values = new ArrayList<>();
        inorder(root, values::add);
        return values;
    }

    private static void inorder(TreeNode root, IntConsumer action) {
        var stack = Collections.asLifoQueue(new ArrayDeque<TreeNode>());

        while (root != null || !stack.isEmpty()) {
            // Go left as far as possible.
            for (; root != null; root = root.left) stack.add(root);

            // Do the inorder action.
            var cur = stack.remove();
            action.accept(cur.val);

            // If we can go right, do that next.
            root = cur.right;
        }
    }
}
